### PR TITLE
fix: properly encode a negative integer in CBOR

### PR DESCRIPTION
**Pull request by etorreborre**
This PR fixes an edge case with the CBOR encoding of `-2^64`. Because of a off-by-one error that value was encoded as a bytestring, even though it fits into a native negative CBOR int.

## Summary by CodeRabbit

* **Bug Fixes**
  * Corrected CBOR encoding and decoding for large positive and negative integers, including the `-2^64` boundary.
  * Improved handling of native integer and bignum representations.
  * Ensured byte strings are encoded consistently, including larger values.

* **Tests**
  * Added coverage for integer boundary values, canonical encodings, and encode/decode round trips.

* **Documentation**
  * Updated the changelog with the CBOR encoding fix.

### CODE REVIEW

File: crates/amaru-uplc/src/flat/data.rs [hunk 1]
```diff
@@ -259,14 +259,22 @@ impl<C> minicbor::encode::Encode<C> for PlutusData<'_> {
...
+                                    .expect("negative integer within CBOR native range");
+                                e.int(integer)?;
+                            }
+                            Err(_) => {
+                                e.tag(Tag::new(3))?;
+                                let (_sign, bytes) = abs_minus_one.to_bytes_be();
+                                encode_bytestring(e, &bytes)?;
+                            }
                         }
                     }
```
Comment (KtorZ) [inline crates/amaru-uplc/src/flat/data.rs]: So, double-checking, this is indeed correct and what the Haskell implementation does: https://hackage-content.haskell.org/package/cborg-0.2.10.0/docs/src/Codec.CBOR.Write.html#local-6989586621679102806

For -2^64, we indeed have two digits (`[0, 1]`) returned by `to_u64_digits()` since the absolute value cannot fit in a single u64. 

It's a bit unfortunate though that we have to do all these conversions and shifting when we use a big int library that does it most entirely for us already. And it's annoying that minicbor does not provide a nice way to encode negative integers directly. 

But, ...perhaps we should consider an implementation/helper in minicbor_extra like: 

```rs
fn encode_bigint<'a, W: encode::Write>(
    e: &'a mut Encoder<W>,
    i: BigInt,
) -> Result<&'a mut Encoder<W>, encode::Error<W::Error>> {
    let (header, tag, bytes) = match i.sign() {
        num_bigint::Sign::NoSign => {
            e.u8(0)?;
            return Ok(());
        }
        num_bigint::Sign::Plus => (0x00, 2, i.to_bytes_be().1),
        num_bigint::Sign::Minus => (0x20, 3, (i.neg() - BigInt::one()).to_bytes_be().1),
    };

    match bytes.len() {
        1 if bytes[0] <= 23 => {
            put(e, &bytes)?;
        }
        len @ 1..=8 => {
            put(e, &[header | (23 + len.bit_width() as u8)])?;
            put(e, &bytes)?;
        }
        _ => {
            e.tag(Tag::new(tag))?;
            encode_bytestring(e, &bytes)?;
        }
    }

    Ok(())
}
```

Where `put` is simply:

```rs
fn put<W: minicbor::encode::Write>(
    e: &mut minicbor::Encoder<W>,
    bytes: &[u8],
) -> Result<(), minicbor::encode::Error<W::Error>> {
    e.writer_mut().write_all(bytes).map_err(minicbor::encode::Error::write)
}
```

This way, we have one clean and relatively effecient bigint implementation which we can rely on not only when encoding data but anywhere we might need bigints...

Comment (etorreborre) [reply crates/amaru-uplc/src/flat/data.rs]: @KtorZ I reworked that part of the code in the direction you indicated and added the relevant tests.

**Review by KtorZ — APPROVED** [review]

**Review by yHSJ — APPROVED** [review]
I checked the Haskell source that Matthias shared, and I am not proficient enough in Haskell to understand it enough to compare to this. So, my approval here is largely based  on his review comment that said the logic is correct and needed.

I am a bit surprised that we don't run into *any* issues with the Plutus conformance tests around negative numbers...

**Comment by Quantumplation** [comment]
@yHSJ that's not suprising to me because:

1) it's only for -2^64, not just all negative values; anything smaller gets correctly encoded, and anything bigger *should* be encoded as a bytestring. This is a boundary condition.
2) *intermediate* results in plutus go negative all the time, but those don't get encoded to cbor very often
3) the *parser* is very permissive; I think the bytestring encoding is still *valid*, just not canonical (Duncan's recurring gift)

So, it would require a conformance test that *exactly*:
- encoded -2^64 in cbor in plutus
- *depended* on that encoding (like hashing it, or checking an ed25519 signature)

That's a pretty easy needle-eye to miss. Would be a great addition though!

**Comment by etorreborre** [comment]
> That's a pretty easy needle-eye to miss. Would be a great addition though!

Yes that should definitely be covered by conformance tests.
